<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>statsintro</title>
    <meta charset="utf-8" />
    <meta name="author" content="Bengt Sennblad" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/metropolis-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# statsintro
### Bengt Sennblad
### 9/30/2020

---

class: inverse, middle, center
&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=800px&gt;&lt;/html&gt; 





# Why Statistics?

---

# Content/aim



.pull-left[
## What do we want you to gain from this course?
- Learn a selection statistical methods


]
---

# Content/aim



.pull-left[
## What do we want you to gain from this course?
- Learn a selection statistical methods

### ...but, more importantly,
- be able to, by yourselves, explore additional methods
  - finding
  - get a intuitive understanding
  - evaluate assumptions
    - know when to apply
    - know when not to apply
]
--
.pull-right[
### How?
- Go from black-box to understanding
    -several levels
      - models = assumptions
      - tests -- test statistics
      - overparmeterizations
- Appreciate theory as a friend
    - needed for understanding assumptions
    - help when evaluating methods
- basis/framework for learning more
    - reading about statistical methods
    - Basic math language
      - not afraid of equations
      - know some basic terminology

]
---
class: inverse, middle, center
&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=800px&gt;&lt;/html&gt; 

# Examples

---

# Example: Mouse knockout experiment

### Setup
Hypothesis: The low density lipoprotein (LDL) receptor gene *LDLR* affects the prevalence of hypercholsterolaemia.

- 10 wildtype (wt) mice
- 10 *LDLR* knockout (KO) mice

Measure plasma concentration of total cholesterol at one time point after feeeding on high fat diet.

???

low-density lipoprotein
---
layout: true
# Example: Mouse knockout experiment
.pull-left[
![](intro2_files/figure-html/mice1-1.png)&lt;!-- --&gt;
]
---
.pull-right[
### Visualize results in a plot

]
---
.pull-right[
### Visualize results in a plot
- Not so informative!
- Let's improve
]
---
layout: false
layout: true
# Example: Mouse knockout experiment

.pull-left[
![](intro2_files/figure-html/mice1a-1.png)&lt;!-- --&gt;
]

---
.pull-right[

### Improved visualization
  - Collect KO and wt separately as columns in a **boxplot**
]

---
.pull-right[

### Improved visualization
  - Collect KO and wt separately as columns in a boxplot
  ![black box warning](assets/blackbox.jpg)
]

---
.pull-right[

### Improved visualization
  - Collect KO and wt separately as columns in a boxplot
  - Visualize **distribution** of sample values using **descriptive statistics**
      - *mean* 
      - *quartiles*
]
---
.pull-right[

### Interpretation

Intuitively, there is a clear difference between KO and wt in this plot...
]
???
- 

-- 
.pull-right[
What if the resuklt had looked different?

]
---
layout: false
# Example: Mouse knockout experiment

.pull-left[
![](intro2_files/figure-html/mice3-1.png)&lt;!-- --&gt;
]
.pull-right[

### Interpretation

...equally clearly, here is no difference between KO and wt...

]
---
layout: true
# Example: Mouse knockout experiment

.pull-left[
![](intro2_files/figure-html/mice4-1.png)&lt;!-- --&gt;
]

---
.pull-right[

### Interpretation

... Now, the difference is less clear

]
---
.pull-right[

### Interpretation

... Now, the difference is less clear

... we get uncertain

]
---
.pull-right[

### Interpretation

... Now, the difference is less clear

... we get uncertain

... Uncertainty best measured as probabilities!

... with probabilities we can perform a **statistical test**
]
---
.pull-right[

### Interpretation

... Now, the difference is less clear

... we get uncertain

... Uncertainty best measured as probabilities!

... with probabilities we can perform a **statistical test**

  ![black box warning](assets/blackbox.jpg)
]

---
.pull-right[
### Statistical tests

#### Model
- Abstractions of reality  probabilities
- Simplifying assumptions
  - Variation approximately `\(N(\mu,\sigma)\)`
    
#### Hypotheses
- H0: `\(\mu\)` and `\(\sigma\)` same for KO and wt 
- H1: `\(\mu\)` might be different 

#### Test
- **P-value** `\((p)\)` =  probability of observation under H0
- **significance level** `\(p\leq \alpha\)`
]

---
layout: false 
layout: true
# Example: Mouse knockout experiment

.pull-left[
![](intro2_files/figure-html/mice5-1.png)&lt;!-- --&gt;
]
---
.pull-right[

### Interpretation

... what about here?.

]


---
.pull-right[

### Interpretation

...what about here?.

#### Tentative answer
This might require testing hypotheses that also look at the *variance*

]


---
layout: false
layout: true
# Example: regression


.pull-left[
![](intro2_files/figure-html/regression2-1.png)&lt;!-- --&gt;
]
---
.pull-right[
Dependencies between variables

- Relation between two variables -- **correlation**
]

???

Hmnm try to find some biological example for this
---
.pull-right[
Dependencies between variables

- Relation between two variables -- **correlation**
  ![black box warning](assets/blackbox.jpg)

]
 
???

Actually, not going to throgh this here .. have to wait until the session on this
---
.pull-right[
Dependencies between variables

- Relation between two variables -- **correlation**

- One variable depends on the other -- **regression**
  - **fit** a **linear model** 
]

???

Need some good biological example

---
.pull-right[
Dependencies between variables

- Relation between two variables -- **correlation**

- One variable depends on the other -- **regression**
  - **fit** a **linear model** 
  ![black box warning](assets/blackbox.jpg)
]
---
layout: false
layout: true
# Example: regression

.pull-left[
![](intro2_files/figure-html/regression3-1.png)&lt;!-- --&gt;
]
---
.pull-right[
## Linear model

- Extend our previous model
  - Variation approximately `\(N(\mu,\sigma)\)`  
  *and*
  - `\(\mu = \alpha +\beta x\)` 
       

]

???
Maybe this can suffice as hierarchical model example instead...?
---
.pull-right[
## Linear model

- Extend our previous model
  - Variation approximately `\(N(\mu,\sigma)\)`  
  *and*
  - `\(\mu = \alpha +\beta x\)` 

- Uses
  1. Simulation
      - validating methods

]
---
.pull-right[
## Linear model

- Extend our previous model
  - Variation approximately `\(N(\mu,\sigma)\)`  
  *and*
  - `\(\mu = \alpha +\beta x\)` 

- Uses
  1. Simulation
      - validating methods
  2. Inference 
      - **regression**
          - optimal `\(\alpha\)` and `\(\beta\)` values


]
???
- `\(\beta\)` tells how strongly y depend on x
- not always this simple
---
layout: false
layout: true
# Example: regression

.pull-left[
![](intro2_files/figure-html/regression4-1.png)&lt;!-- --&gt;
]

---
.pull-right[
## Linear model

- Extend our previous model
  - Variation approximately `\(N(\mu,\sigma)\)`  
  *and*
  - `\(\mu = \alpha +\beta x\)` 

- Uses
  1. Simulation
      - validating methods
  2. Inference 
      - **regression**
          - optimal `\(\alpha\)` and `\(\beta\)` values


]

???

- much higher `\(\sigma=10\)` (before `\(\sigma=1\)`)
---
layout: false
layout: true
# Example: regression

.pull-left[
![](intro2_files/figure-html/regression5-1.png)&lt;!-- --&gt;
]
---
.pull-right[
... What about here


]
---
.pull-right[
... What about here

## Non-linear dependencies

]
---
.pull-right[
... What about here

## Non-linear dependencies

- **Generalized linear models (GLMs)**
  ![black box warning](assets/blackbox.jpg)

]
---
layout: false
# Hierarchical models

.pull-left[
![](intro2_files/figure-html/hierarchical1-1.png)&lt;!-- --&gt;
]
.pull-right[

Linear model can be viewed as a hierarchical model

- `\(y \sim N(\mu, \sigma)\)`
    - `\(\mu = \beta x\)`

]

#### [I should try to find  better example for overparmeterization!? -- and skip his example
  - fits beeter with multivariate?
  - Although this is neat in some ways (but not all)]
---
# Hierarchical models

.pull-left[
![](intro2_files/figure-html/hierarchical2-1.png)&lt;!-- --&gt;
]
.pull-right[

What is this?

... could it be related to gender


]
---
# Hierarchical models

.pull-left[
![](intro2_files/figure-html/hierarchical3-1.png)&lt;!-- --&gt;
]
.pull-right[

What is this?

... could it be related to gender

(This is not the most common solution, but let's play with hierarchical models)
]
---
layout: true
# Hierarchical models

.pull-left[
![](intro2_files/figure-html/hierarchical4-1.png)&lt;!-- --&gt;
]
---
.pull-right[

- Extended hierarchical model
  - `\(y \sim N(\mu, \sigma)\)`
  - `\(\mu = \beta x\)`
  - `\(\beta = \begin{cases} \beta_f &amp; \textrm{for females}\\ \beta_m &amp; \textrm{for males}\end{cases}\)`

... Let's play some more...  

]

???

- Note! To simplify, I've used normalized valies so `\(\alpha=0\)`
---
.pull-right[

- Extended hierarchical model
  - `\(y \sim N(\mu, \sigma)\)`
  - `\(\mu = \beta x\)`
  - `\(\beta = \begin{cases} \beta_f &amp; \textrm{for females}\\ \beta_m &amp; \textrm{for males}\end{cases}\)`
  - `\(\sigma = \begin{cases} \sigma_f &amp; \textrm{for females}\\ \sigma_m &amp; \textrm{for males}\end{cases}\)`


]
---
layout: false
layout: true
# Over-parameterization

.pull-left[
![](intro2_files/figure-html/hierarchical5-1.png)&lt;!-- --&gt;
]

---

.pull-right[

- Extended hierarchical model
  - `\(y \sim N(\mu, \sigma)\)`
  - `\(\mu = \beta x\)`
  - `\(\beta = \begin{cases} \beta_f &amp; \textrm{for females}\\ \beta_m &amp; \textrm{for males}\end{cases}\)`
  - `\(\sigma = \begin{cases} \sigma_f &amp; \textrm{for females}\\ \sigma_m &amp; \textrm{for males}\end{cases}\)`

]

---

.pull-right[

- Extended hierarchical model
  - `\(y \sim N(\mu, \sigma)\)`
  - `\(\mu = \beta x\)`
  - `\(\beta = \begin{cases} \beta_f &amp; \textrm{for females}\\ \beta_m &amp; \textrm{for males}\end{cases}\)`
  - `\(\sigma = \begin{cases} \sigma_f &amp; \textrm{for females}\\ \sigma_m &amp; \textrm{for males}\end{cases}\)`
- Several solution become equally probable! Too many parameters to estimate! **over-parameterization**
- Solution is to limit the parameter space by **regularization**, **feature selection** or  *Bayesian statistics*

]
---

.pull-right[
- Extended hierarchical model
  - `\(y \sim N(\mu, \sigma)\)`
  - `\(\mu = \beta x\)`
  - `\(\beta = \begin{cases} \beta_f &amp; \textrm{for females}\\ \beta_m &amp; \textrm{for males}\end{cases}\)`
  - `\(\sigma = \begin{cases} \sigma_f &amp; \textrm{for females}\\ \sigma_m &amp; \textrm{for males}\end{cases}\)`
- Several solution become equally probable! Too many parameters to estimate! **over-parameterization**
- Solution is to limit the parameter space by **regularization**, **feature selection** or  *Bayesian statistics*

  ![black box warning](assets/blackbox.jpg)
]

???

- Leave this black box unexplained for the time being, see course:)

---
layout: false
layout: true
# Example: GWAS
.pull-left[
![](intro2_files/figure-html/gwas-1.png)&lt;!-- --&gt;

]
---
.pull-right[

### Manhattan plot

- Individual association test for each SNP and disease
- `\(-log_{10}\)` P-value against genomic position

]

???
-log(0.05) = 1.30103
- thousands of associations
---

.pull-right[

### Manhattan plot

- Individual association test for each SNP and disease
- `\(-log_{10}\)` P-value against genomic position

#### *Multiple tests*
- Repeated experiments are more likely to succeed under H0


]

???

- Discuss repeated dice throws
- This manhattan plot is actually generated under the null model
---
.pull-right[

### Manhattan plot

- Individual association test for each SNP and disease
- `\(-log_{10}\)` P-value against genomic position

#### *Multiple tests*
- Repeated experiments are more likely to succeed under H0

#### Multiple test correction
- **Bonferroni, False Discovery Rate (FDR)**

]

---
.pull-right[

### Manhattan plot

- Individual association test for each SNP and disease
- `\(-log_{10}\)` P-value against genomic position

#### *Multiple tests*
- Repeated experiments are more likely to succeed under H0

#### Multiple test correction
- **Bonferroni, False Discovery Rate (FDR)**
  ![black box warning](assets/blackbox.jpg)

]


---
layout: false
layout: true
# Multivariate analysis

.pull-left[
![](intro2_files/figure-html/multivariate-1.png)&lt;!-- --&gt;
]
---
.pull-right[
## Multivariate regression
- We can visualize models with 3 variables
- and write an equation for this:  
`\(\begin{cases}y \sim N(\mu, \sigma) \\\mu = \alpha+\beta x + \gamma gender\end{cases}\)`

 
]

---
.pull-right[
## Multivariate regression

 
]

---
.pull-right[
## Multivariate regression

 ![black box warning](assets/blackbox.jpg) 
]

---
.pull-right[
## Multivariate regression

- We can visualize models with 3 variables
- and write an equation for this:  
`\(\begin{cases}y \sim N(\mu, \sigma) \\\mu = \alpha+\beta x + \gamma gender\end{cases}\)`
- For &gt;3 variables, visualization breaks...
- but we can still write it as an equation:  
`\(\begin{cases}y \sim N(\mu, \sigma) \\\mu = \alpha+\beta x + \gamma gender + \delta age\end{cases}\)`


]
---
.pull-right[
## Multivariate regression

- We can visualize models with 3 variables
- and write an equation for this:  
`\(\begin{cases}y \sim N(\mu, \sigma) \\\mu = \alpha+\beta x + \gamma gender\end{cases}\)`
- For &gt;3 variables, visualization breaks...
- but we can still write it as an equation:  
`\(\begin{cases}y \sim N(\mu, \sigma) \\\mu = \alpha+\beta x + \gamma gender + \delta age\end{cases}\)`
]
---
.pull-right[
## Multivariate regression

- We can visualize models with 3 variables
- and write an equation for this:  
`\(\begin{cases}y \sim N(\mu, \sigma) \\\mu = \alpha+\beta x + \gamma gender\end{cases}\)`

 
- For &gt;3 variables, visualization breaks...
- but we can still write it as an equation:  
`\(\begin{cases}y \sim N(\mu, \sigma) \\\mu = \alpha+\beta x + \gamma gender + \delta age\end{cases}\)`

- Big Data `\(\Rightarrow\)` risk for over-parameterization `\(\Rightarrow\)` feature selection [Better here?]

]

---
layout: false
layout: true
# Big data - Machine learning

## PCA and clustering

.pull-left[
![](intro2_files/figure-html/pca-1.png)&lt;!-- --&gt;
]
---
.pull-right[

### Feature selection with PCA

- **linear transformation** of original variables into new *hidden variables*, components/factors

### Clustering
- identify groups based on hidden variables
]
---
.pull-right[

### Feature selection with PCA

- **linear transformation** of original variables into new *hidden variables*, components/factors

### Clustering
- identify groups based on hidden variables
  ![black box warning](assets/blackbox.jpg)

]
---
layout: true

# Big data - Machine learning



## Artificial neural network (ANN)

- **Train** ANN to learn which variables are important and how to combine them to predict the right answer.


---
.pull-left[
&lt;br&gt;&lt;br&gt;
&lt;p style="text-align: center; font-size:2em;"&gt;Some Big Data 
multivariate input!
&lt;/p&gt;
].pull-right[
![](intro2_files/figure-html/ann2-1.png)&lt;!-- --&gt;
]
---
.pull-left[
##### Input  could be, e.g.,  images
&lt;img src="assets/Nora.jpeg" width="200" style="display: block; margin: auto 0 auto auto;" /&gt;
].pull-right[
![](intro2_files/figure-html/ann4-1.png)&lt;!-- --&gt;
]
---
.pull-left[
#####  ... or Input could be genome sequences...
...AGGGGCATAAAAGGTTCTACGCGCAAA
TCTTACTTTTGGAAACTCAGATGCCGGACT
TCAGTGTCCCCCAATATCCGAATACGGCTC
GGGCTTTCCACTGCGCGTCCTCGACTAAAT
TTTAAGCTAGAACGGAATCTACACTCAAGC
GGACGGAGCTGTCTTACTAAAGTCGATGCA
AAAGTCGTGGCTATGTCAATGGCCCAGAGC
ACGCAGTCCCTGCCATCGCATTAGTTTTTT
CGAGTCGTCAACCGCGCGTCAGCCTTC...
]
.pull-right[
![](intro2_files/figure-html/ann6-1.png)&lt;!-- --&gt;
]

---
.pull-left[
#####  This is of course a major...
![black box warning](assets/blackbox.jpg)
]
.pull-right[
![](intro2_files/figure-html/ann8-1.png)&lt;!-- --&gt;

]



 
 
???
This is of course a major black box


---
layout: false
# Bayesian statistics example: phylogeny



.pull-left[
![](intro2_files/figure-html/phylogeny2-1.png)&lt;!-- --&gt;
]
.pull-right[
## Clustering methods
- works with pairwise distances
- fast
- single solution

### Problem 
- different methods yield different solutions


---
# Bayesian statistics example: phylogeny

.pull-left[
![](intro2_files/figure-html/phylogeny3-1.png)&lt;!-- --&gt;
]
.pull-right[
## Robust?
- Using branch lengths as a measure, there is little support for either placement
]
---
layout: true
# Bayesian statistics example: phylogeny

.pull-left[
![](intro2_files/figure-html/phylogeny4-1.png)&lt;!-- --&gt;
]
---
.pull-right[
## In fact there are multiple possible trees

- Is one better than the other?
]


---
.pull-right[
## In fact there are multiple possible trees

- Is one better than the other?

### *Frequentist approach*
  - hypothesis test
  
  

]

---
.pull-right[
## In fact there are multiple possible trees

- Is one better than the other?

### *Frequentist approach*
  - hypothesis test
  
### *Bayesian approach*
  - assign *posterior probabilities* of each tree given data
  - Look at the whole distribution
  - evaluate statistic of interest
  

]
---
.pull-right[
### *Bayesian approach**
- Here, statistic of interest = Support for individual clades (*clade credibility*, CC)
  - CC of clade a = sum of trees with clade A weighted by their probabilities
  
![](intro2_files/figure-html/phylo5-1.png)&lt;!-- --&gt;

  ]
--

- Note. Frequentist approach to clade support: use data resampling (bootstrap)

---
layout: false

# "*Frequentist*" vs Bayesian statstics

[Note! maybe include this in reguilarization lecture instead? -- check if doable]


.pull-left[
### Frequentist
- P-values
  - probability of data given parameters (likelihood)
  - "pairwise" hypothesis tests 

- "Point estimates" focus?

- Hierarchical models
  - regularization

- confidence interval

- **always check model assumptions!**

]
.pull-right[
### Bayesian
- Posterior probabilities
  - probability of parameters given data
  - posterior distributions

- Sum over posterior distributions

- Hierarchical priors
   - prior distribution constraints

- posterior interval

- **always check model assumptions!**

]

&lt;!---
---

# POssible Bayesian example
.pull-left[
#```{r, bayes1}
#p9
#```
#]
.pull-right[

Complete hierarchical model
- `\(y \sim N(\mu, \sigma)\)`
- `\(\mu = \beta x\)`
- `\(\beta = \begin{cases} \beta_m &amp; \textrm{for males}\\ \beta_f &amp; \textrm{for males}\end{cases}\)`
- `\(\sigma = \begin{cases} \sigma_m &amp; \textrm{for males}\\ \sigma_f &amp; \textrm{for males}\end{cases}\)`

prior probabilities using prior knowledge
- `\(\begin{cases}\beta_m \sim N(\beta_f, 0.05) \\ \sigma_m \sim N(\sigma_f, 0.0001)\end{cases}\)`
- `\(\begin{cases} \beta_f \sim N(0, 0.5)\\ \sigma_f \sim N(0, 0.5) \end{cases}\)`

*posterior probability* distributions of the parameter values given the data



]

---&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "<div class=\"progress-bar-container\">   <div class=\"progress-bar\" style=\"width: calc(%current% / %total% * 100%);\">   </div> </div>"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
